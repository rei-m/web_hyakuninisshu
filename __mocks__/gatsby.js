const React = require("react");
const { action } = require("@storybook/addon-actions");
const iconImage = require("@src/images/app-icon.png");
const correctImage = require("@src/images/check_correct.png");
const incorrectImage = require("@src/images/check_incorrect.png");
const dogezaImage = require("@src/images/dogeza_businessman.png");
const karutaImage = require("@src/images/karuta_001.jpg");
const tatamiImage  = require("@src/images/tatami_part.png");

const linkActionHandler = action("Link:");
const navigateActionHandler = action("NavigateTo:");

const imageSrc = (image) => (image.default === "test-file-stub" ? image.default : image);

module.exports = {
  graphql: (args) => args,
  Link: ({ to, ...rest }) =>
    React.createElement("a", {
      ...rest,
      href: to,
      onClick: (e) => {
        e.preventDefault();
        linkActionHandler(to);
      }
    }),
  StaticQuery: () => {},
  useStaticQuery: (args) => {
    const query = args[0];
    if (query.indexOf("query SEOQuery") > -1) {
      return {
        site: {
          siteMetadata: {
            title: "百人一首 簡単に暗記",
            description: "説明",
            author: "rei-m",
          },
        },
        ogpImage: {
          publicURL: imageSrc(iconImage),
        },
      };
    }
    if (query.indexOf("query CorrectImageQuery") > -1) {
      return {
        correctImage: {
          childImageSharp: {
            fluid: {
              aspectRatio: 0.7228915662650602,
              srcSet: "",
              sizes: "(max-width: 300px) 100vw, 200px",
              src: imageSrc(correctImage),
            },
          },
        },
        incorrectImage: {
          childImageSharp: {
            fluid: {
              aspectRatio: 0.7228915662650602,
              srcSet: "",
              sizes: "(max-width: 300px) 100vw, 200px",
              src: imageSrc(incorrectImage),
            },
          },
        },
      };
    }
    if (query.indexOf("query DogezaImageQuery") > -1) {
      return {
        dogezaImage: {
          childImageSharp: {
            fluid: {
              aspectRatio: 0.7228915662650602,
              srcSet: "",
              sizes: "(max-width: 200px) 200px, 200px",
              src: imageSrc(dogezaImage),
            },
          },
        },
      };
    }
    if (query.indexOf("query TatamiImageQuery") > -1) {
      return {
        tatamiImage: {
          publicURL: tatamiImage,
        },
      };
    }
    if (query.indexOf("query KarutaImageQuery") > -1) {
      return {
        karutaImages: {
          edges: Array.from(Array(100).keys()).map(i => ({
            node: {
              childImageSharp: {
                fluid: {
                  aspectRatio: 0.7228915662650602,
                  srcSet: "",
                  sizes: "(max-width: 200px) 100vw, 200px",
                  src: imageSrc(karutaImage),
                }
              },
              name: `karuta_${("00" + (i + 1)).slice(-3)}`
            }
          }))
        }
      };
    }
    return {};
  },
  navigate: (to, options) => {
    navigateActionHandler(to, options);
  }
}
