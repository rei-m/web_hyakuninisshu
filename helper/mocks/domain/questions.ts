import { Question } from '@src/domain/models';

export const MOCK_QUESTION_1: Question = {
  id: 1,
  correctAnswerKarutaNo: 1,
  choiceKarutaNoList: [1, 2, 3, 4],
};

export const MOCK_QUESTION_2: Question = {
  id: 2,
  correctAnswerKarutaNo: 2,
  choiceKarutaNoList: [1, 2, 3, 4],
};

export const MOCK_QUESTION_1_STARTED: Question = {
  id: 1,
  correctAnswerKarutaNo: 1,
  choiceKarutaNoList: [1, 2, 3, 4],
  startTime: 1583595557624,
};

export const MOCK_QUESTION_1_ANSWERED_CORRECT: Question = {
  id: 1,
  correctAnswerKarutaNo: 1,
  choiceKarutaNoList: [1, 2, 3, 4],
  startTime: 1583595557624,
  answer: {
    isCorrect: true,
    answerTime: 1583595559624,
  },
};

export const MOCK_QUESTION_2_ANSWERED_WRONG: Question = {
  id: 2,
  correctAnswerKarutaNo: 2,
  choiceKarutaNoList: [1, 2, 3, 4],
  startTime: 1583595689123,
  answer: {
    isCorrect: false,
    answerTime: 1583595690123,
  },
};
