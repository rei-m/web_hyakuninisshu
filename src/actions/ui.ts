import { Action, Dispatch } from 'redux';
import { Color, Kimariji } from '@src/types';

export const OPEN_KARUTAS_FILTER_NAME = 'OPEN_KARUTAS_FILTER';
export type OPEN_KARUTAS_FILTER_TYPE = typeof OPEN_KARUTAS_FILTER_NAME;

export const CLOSE_KARUTAS_FILTER_NAME = 'CLOSE_KARUTAS_FILTER';
export type CLOSE_KARUTAS_FILTER_TYPE = typeof CLOSE_KARUTAS_FILTER_NAME;

export const TOGGLE_KARUTAS_COLOR_NAME = 'TOGGLE_KARUTAS_COLOR';
export type TOGGLE_KARUTAS_COLOR_TYPE = typeof TOGGLE_KARUTAS_COLOR_NAME;

export const TOGGLE_KARUTAS_KIMARIJI_NAME = 'TOGGLE_KARUTAS_KIMARIJI_NAME';
export type TOGGLE_KARUTAS_KIMARIJI_TYPE = typeof TOGGLE_KARUTAS_KIMARIJI_NAME;

export interface OpenKarutasFilterAction extends Action {
  readonly type: OPEN_KARUTAS_FILTER_TYPE;
}

export interface CloseKarutasFilterAction extends Action {
  readonly type: CLOSE_KARUTAS_FILTER_TYPE;
}

export interface OKarutasFilterAction extends Action {
  readonly type: OPEN_KARUTAS_FILTER_TYPE;
}

export interface ToggleKarutasColorAction extends Action {
  readonly type: TOGGLE_KARUTAS_COLOR_TYPE;
  readonly payload: {
    readonly color: Color;
    readonly checked: boolean;
  };
}

export interface ToggleKarutasKimarijiAction extends Action {
  readonly type: TOGGLE_KARUTAS_KIMARIJI_TYPE;
  readonly payload: {
    readonly kimariji: Kimariji;
    readonly checked: boolean;
  };
}

export type UiActions =
  | OpenKarutasFilterAction
  | CloseKarutasFilterAction
  | ToggleKarutasColorAction
  | ToggleKarutasKimarijiAction;

/*
 * action creators
 */
export const openKarutasFilter = () => (dispatch: Dispatch<OpenKarutasFilterAction>) =>
  dispatch({
    type: OPEN_KARUTAS_FILTER_NAME,
  });

export const closeKarutasFilter = () => (dispatch: Dispatch<CloseKarutasFilterAction>) =>
  dispatch({
    type: CLOSE_KARUTAS_FILTER_NAME,
  });

export const toggleKarutasColor = (color: Color, checked: boolean) => (
  dispatch: Dispatch<ToggleKarutasColorAction>
) => {
  const action: ToggleKarutasColorAction = {
    payload: {
      color,
      checked,
    },
    type: TOGGLE_KARUTAS_COLOR_NAME,
  };

  dispatch(action);
};

export const toggleKarutasKimariji = (kimariji: Kimariji, checked: boolean) => (
  dispatch: Dispatch<ToggleKarutasKimarijiAction>
) => {
  const action: ToggleKarutasKimarijiAction = {
    payload: {
      kimariji,
      checked,
    },
    type: TOGGLE_KARUTAS_KIMARIJI_NAME,
  };

  dispatch(action);
};
