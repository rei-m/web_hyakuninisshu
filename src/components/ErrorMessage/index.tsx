import * as React from 'react';
import styled from '@src/styles/styled-components';

export interface Props {
  text: string;
}

const Container = styled.div`
  height: 300px;
  display: flex;
  justify-content: center;
  align-items: center;
`;

const Message = styled.p`
  font-size: 1.6rem;
`;

const ErrorMessage: React.FC<Props> = ({ text }) => (
  <Container>
    <Message>{text}</Message>
  </Container>
);

export default ErrorMessage;
