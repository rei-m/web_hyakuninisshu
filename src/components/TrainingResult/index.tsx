import * as React from 'react';
import { Link } from 'react-router-dom';
// import styled from 'styled-components';

export interface TrainingResultOwnProps {
  readonly totalCount: number;
  readonly correctCount: number;
  readonly averageAnswerSecond: number;
}

export interface TrainingResultDispatchProps {
  readonly onClickRestart: () => void;
}

export type TrainingResultProps = TrainingResultOwnProps &
  TrainingResultDispatchProps;

const TrainingResult = ({
  averageAnswerSecond,
  correctCount,
  totalCount,
  onClickRestart
}: TrainingResultProps) => {
  return (
    <section>
      <div>
        <div>正解数</div>
        <div>
          {correctCount} / {totalCount}
        </div>
      </div>
      <div>
        <div>平均回答時間</div>
        <div>{averageAnswerSecond}秒</div>
      </div>
      {correctCount !== totalCount && (
        <button onClick={onClickRestart}>間違えた歌の練習</button>
      )}
      <Link to="/training" replace={true}>
        メニューに戻る
      </Link>
    </section>
  );
};

export default TrainingResult;
