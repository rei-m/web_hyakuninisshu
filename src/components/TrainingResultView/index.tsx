import * as React from 'react';
import { navigate } from 'gatsby';
import styled from '@src/styles/styled-components';
import { appTheme } from '@src/styles/theme';
import QuestionResultsSummary from '@src/components/QuestionResultsSummary';
import AppButton from '@src/components/AppButton';
import TweetButton from '@src/components/TweetButton';

export interface Props {
  totalCount: number;
  correctCount: number;
  averageAnswerSecond: number;
  onClickRestart: () => void;
}

const Container = styled.div`
  max-width: 380px;
  padding: ${({ theme }) => theme.spacing2x};
  margin: auto;
`;

const ShareButtonsWrapper = styled.div`
  text-align: center;
`;

const onClickBack = () => {
  navigate('/training', { replace: true });
};

const QuestionsResult: React.FC<Props> = ({ averageAnswerSecond, correctCount, totalCount, onClickRestart }) => (
  <Container>
    <QuestionResultsSummary
      title="正解数"
      value={`${correctCount} / ${totalCount}`}
      style={{ marginBottom: appTheme.spacing2x }}
    />
    <QuestionResultsSummary
      title="平均回答時間"
      value={`${averageAnswerSecond}秒`}
      style={{ marginBottom: appTheme.spacing4x }}
    />
    <ShareButtonsWrapper>
      <TweetButton text={`百人一首で ${totalCount}問中 ${correctCount}問 正解しました！`} />
    </ShareButtonsWrapper>
    {correctCount !== totalCount && (
      <AppButton
        label="間違えた歌の練習をする"
        icon="refresh"
        type="normal"
        onClick={onClickRestart}
        data-test="restart"
        style={{ margin: appTheme.spacing2x }}
      />
    )}
    <AppButton label="メニューに戻る" icon="arrow_back" type="normal" onClick={onClickBack} data-test="back" />
  </Container>
);

export default QuestionsResult;
