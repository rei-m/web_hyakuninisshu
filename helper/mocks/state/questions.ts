import { questionsTypes } from '@src/state/questions';

export const MOCK_YOMIFUDA_1: questionsTypes.YomiFuda = {
  karutaNo: 1,
  shoku: '秋の田の',
  niku: 'かりほの庵の',
  sanku: '苫をあらみ',
};

export const MOCK_TORIFUDA_1: questionsTypes.ToriFuda = {
  karutaNo: 1,
  shiku: 'きりにぬれつつ',
  kekku: 'わがころもでは',
};

export const MOCK_TORIFUDA_2: questionsTypes.ToriFuda = {
  karutaNo: 2,
  shiku: 'ころもほすてふ',
  kekku: 'あまのかぐやま',
};

export const MOCK_TORIFUDA_3: questionsTypes.ToriFuda = {
  karutaNo: 3,
  shiku: 'ながながしよを',
  kekku: 'ひとりかもねむ',
};

export const MOCK_TORIFUDA_4: questionsTypes.ToriFuda = {
  karutaNo: 4,
  shiku: 'ふじのたかねに',
  kekku: 'ゆきはふりつつ',
};
