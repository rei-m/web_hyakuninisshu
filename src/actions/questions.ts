import { Action, Dispatch } from 'redux';
import { GlobalState } from '@src/state';
import { Answer, Karuta, Question } from '@src/types';
import { randomizeArray } from '@src/utils';
import { fetchTorifudas, questionsFilter, toDulation } from '@src/utils/questions';
import {
  ColorCondition,
  KarutaStyleCondition,
  KimarijiCondition,
  QuestionAnimCondition,
  QuestionState,
  RangeFromCondition,
  RangeToCondition,
} from '@src/enums';

export const START_TRAINING_NAME = 'START_TRAINING_NAME';
export type START_TRAINING_TYPE = typeof START_TRAINING_NAME;

export const START_EXAM_NAME = 'START_EXAM_NAME';
export type START_EXAM_TYPE = typeof START_EXAM_NAME;

export const RESTART_QUESTIONS_NAME = 'RESTART_QUESTIONS_NAME';
export type RESTART_QUESTIONS_TYPE = typeof RESTART_QUESTIONS_NAME;

export const ANSWER_QUESTION_NAME = 'ANSWER_QUESTION_NAME';
export type ANSWER_QUESTION_TYPE = typeof ANSWER_QUESTION_NAME;

export const CONFIRM_CORRECT_NAME = 'CONFIRM_CORRECT_NAME';
export type CONFIRM_CORRECT_TYPE = typeof CONFIRM_CORRECT_NAME;

export const OPEN_NEXT_QUESTION_NAME = 'OPEN_NEXT_QUESTION_NAME';
export type OPEN_NEXT_QUESTION_TYPE = typeof OPEN_NEXT_QUESTION_NAME;

export const FINISH_QUESTIONS_NAME = 'FINISH_QUESTIONS_NAME';
export type FINISH_QUESTIONS_TYPE = typeof FINISH_QUESTIONS_NAME;

export interface StartTrainingAction extends Action {
  readonly type: START_TRAINING_TYPE;
  readonly payload: {
    readonly karutas: Karuta[];
    readonly questions: Question[];
    readonly startedTime: number;
    readonly nextState: QuestionState;
    readonly dulation: number;
  };
  readonly meta: {
    readonly rangeFrom: RangeFromCondition;
    readonly rangeTo: RangeToCondition;
    readonly kimariji: KimarijiCondition;
    readonly color: ColorCondition;
    readonly kamiNoKuStyle: KarutaStyleCondition;
    readonly shimoNoKuStyle: KarutaStyleCondition;
    readonly questionAnim: QuestionAnimCondition;
  };
}

export interface StartExamAction extends Action {
  readonly type: START_EXAM_TYPE;
  readonly payload: {
    readonly karutas: Karuta[];
    readonly questions: Question[];
    readonly startedTime: number;
    readonly nextState: QuestionState;
  };
}

export interface RestartQuestionsAction extends Action {
  readonly type: RESTART_QUESTIONS_TYPE;
  readonly payload: {
    readonly questions: Question[];
    readonly startedTime: number;
    readonly nextState: QuestionState;
  };
}

export interface AnswerQuestionAction extends Action {
  readonly type: ANSWER_QUESTION_TYPE;
  readonly payload: {
    readonly answer: Answer;
    readonly nextState: QuestionState;
  };
}

export interface ConfirmCorrectAction extends Action {
  readonly type: CONFIRM_CORRECT_TYPE;
  readonly payload: {
    readonly nextState: QuestionState;
  };
}

export interface OpenNextQuestionAction extends Action {
  readonly type: OPEN_NEXT_QUESTION_TYPE;
  readonly payload: {
    readonly nextIndex: number;
    readonly nextState: QuestionState;
    readonly startedTime: number;
  };
}

export interface FinishQuestionsAction extends Action {
  readonly type: FINISH_QUESTIONS_TYPE;
  readonly payload: {
    readonly nextState: QuestionState;
  };
}

export type QuestionsActions =
  | StartTrainingAction
  | StartExamAction
  | RestartQuestionsAction
  | AnswerQuestionAction
  | ConfirmCorrectAction
  | OpenNextQuestionAction
  | FinishQuestionsAction;

/*
 * action creators
 */
export const startTraining = (
  karutas: Karuta[],
  rangeFrom: RangeFromCondition,
  rangeTo: RangeToCondition,
  kimariji: KimarijiCondition,
  color: ColorCondition,
  kamiNoKuStyle: KarutaStyleCondition,
  shimoNoKuStyle: KarutaStyleCondition,
  questionAnim: QuestionAnimCondition
) => (dispatch: Dispatch<QuestionsActions>) => {
  const questions = new QuestionsFactory(karutas)
    .setRange(rangeFrom, rangeTo)
    .setKimariji(kimariji)
    .setColor(color)
    .setKamiNoKuStyle(kamiNoKuStyle)
    .setShimoNoKuStyle(shimoNoKuStyle)
    .create();

  const action: StartTrainingAction = {
    meta: {
      color,
      kamiNoKuStyle,
      kimariji,
      rangeFrom,
      rangeTo,
      shimoNoKuStyle,
      questionAnim,
    },
    payload: {
      karutas,
      nextState: QuestionState.InAnswer,
      questions: randomizeArray<Question>(questions),
      startedTime: new Date().getTime(),
      dulation: toDulation(questionAnim),
    },
    type: START_TRAINING_NAME,
  };

  dispatch(action);
};

export const startExam = (karutas: Karuta[]) => (dispatch: Dispatch<QuestionsActions>) => {
  const questions = new QuestionsFactory(karutas)
    .setRange(RangeFromCondition.One, RangeToCondition.OneHundred)
    .setKimariji(KimarijiCondition.None)
    .setColor(ColorCondition.None)
    .setKamiNoKuStyle(KarutaStyleCondition.KanjiAndKana)
    .setShimoNoKuStyle(KarutaStyleCondition.KanaOnly)
    .create();

  const action: StartExamAction = {
    payload: {
      karutas,
      nextState: QuestionState.InAnswer,
      questions: randomizeArray(questions),
      startedTime: new Date().getTime(),
    },
    type: START_EXAM_NAME,
  };

  dispatch(action);
};

export const restartQuestions = () => (dispatch: Dispatch<QuestionsActions>, getState: () => GlobalState) => {
  const { questions, answers } = getState().questions;
  // TODO: 全て回答済みでなかったらエラー
  const finder: { [questionId: number]: Question } = questions.reduce((previous, current) => {
    return { ...previous, [current.id]: current };
  }, {});
  const targets = answers
    .filter(a => !a.correct)
    .map(a => finder[a.questionId])
    .map(q => ({ ...q, toriFudas: randomizeArray(q.toriFudas) }));

  const action: RestartQuestionsAction = {
    payload: {
      nextState: QuestionState.InAnswer,
      questions: randomizeArray(targets),
      startedTime: new Date().getTime(),
    },
    type: RESTART_QUESTIONS_NAME,
  };

  dispatch(action);
};

export const answerQuestion = (questionId: number, karutaNo: number) => (
  dispatch: Dispatch<QuestionsActions>,
  getState: () => GlobalState
) => {
  const { questions, lastStartedTime } = getState().questions;
  const question = questions.find(q => q.id === questionId)!;
  // TODO: QuestionとlastStartedTimeがundefinedの場合
  const correct = question.correctKaruta.no === karutaNo;
  const time = new Date().getTime() - lastStartedTime!;
  const answer = {
    correct,
    karutaNo,
    questionId,
    time,
  };
  const action: AnswerQuestionAction = {
    payload: {
      answer,
      nextState: QuestionState.Answered,
    },
    type: ANSWER_QUESTION_NAME,
  };

  dispatch(action);
};

export const confirmCorrect = (): ConfirmCorrectAction => {
  return {
    payload: {
      nextState: QuestionState.ConfirmCorrect,
    },
    type: CONFIRM_CORRECT_NAME,
  };
};

export const openNextQuestion = () => (dispatch: Dispatch<QuestionsActions>, getState: () => GlobalState) => {
  const { currentIndex } = getState().questions;
  const nextIndex = currentIndex + 1;
  const action: OpenNextQuestionAction = {
    payload: {
      nextIndex,
      nextState: QuestionState.InAnswer,
      startedTime: new Date().getTime(),
    },
    type: OPEN_NEXT_QUESTION_NAME,
  };
  dispatch(action);
};

export const finishQuestions = (): FinishQuestionsAction => {
  return {
    payload: {
      nextState: QuestionState.Finished,
    },
    type: FINISH_QUESTIONS_NAME,
  };
};

class QuestionsFactory {
  private karutas: Karuta[];
  private rangeFrom: number;
  private rangeTo: number;
  private kimariji: number;
  private color: string;
  private kamiNoKuStyle: number;
  private shimoNoKuStyle: number;

  constructor(karutas: Karuta[]) {
    this.karutas = karutas;
    this.rangeFrom = RangeFromCondition.One;
    this.rangeTo = RangeToCondition.OneHundred;
    this.kimariji = KimarijiCondition.None;
    this.color = ColorCondition.None;
    this.kamiNoKuStyle = KarutaStyleCondition.KanjiAndKana;
    this.shimoNoKuStyle = KarutaStyleCondition.KanaOnly;
  }

  public setRange(rangeFrom: number, rangeTo: number) {
    this.rangeFrom = rangeFrom;
    this.rangeTo = rangeTo;
    return this;
  }

  public setKimariji(kimariji: number) {
    this.kimariji = kimariji;
    return this;
  }

  public setColor(color: string) {
    this.color = color;
    return this;
  }

  public setKamiNoKuStyle(kamiNoKuStyle: number) {
    this.kamiNoKuStyle = kamiNoKuStyle;
    return this;
  }

  public setShimoNoKuStyle(shimoNoKuStyle: number) {
    this.shimoNoKuStyle = shimoNoKuStyle;
    return this;
  }

  public create() {
    const targetKarutas = questionsFilter(this.karutas)(this.rangeFrom, this.rangeTo)(this.kimariji)(this.color);

    return targetKarutas.map((k, i) => {
      const id = i + 1;
      const correctKaruta = k;
      const yomiFuda =
        this.kamiNoKuStyle === 0
          ? {
              firstText: k.firstKanji,
              karutaNo: k.no,
              questionId: id,
              secondText: k.secondKanji,
              thirdText: k.thirdKanji,
            }
          : {
              firstText: k.firstKana,
              karutaNo: k.no,
              questionId: id,
              secondText: k.secondKana,
              thirdText: k.thirdKana,
            };
      const toriFudas = fetchTorifudas(this.karutas, correctKaruta).map(toriFuda => {
        return this.shimoNoKuStyle === 0
          ? {
              fifthText: toriFuda.fifthKanji,
              fourthText: toriFuda.fourthKanji,
              karutaNo: toriFuda.no,
              questionId: id,
            }
          : {
              fifthText: toriFuda.fifthKana,
              fourthText: toriFuda.fourthKana,
              karutaNo: toriFuda.no,
              questionId: id,
            };
      });
      return { id, correctKaruta, yomiFuda, toriFudas };
    });
  }
}
