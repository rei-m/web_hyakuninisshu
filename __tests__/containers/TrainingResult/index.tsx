import { GlobalState } from '@src/state';
import * as React from 'react';
import { navigate } from 'gatsby';
import { mapStateToProps, Props, TrainingResult } from '@src/containers/TrainingResult';
import TrainingResultView from '@src/components/TrainingResultView';
import ErrorMessage from '@src/components/ErrorMessage';
import { shallow, ShallowWrapper } from 'enzyme';
import { initialState as questionsState } from '@src/state/questions';
import { initialState as uiState } from '@src/state/ui';
import { QuestionState } from '@src/enums';
import { Answer } from '@src/types';
import { create } from '@test/factories';

describe('<TrainingResult />', () => {
  describe('components', () => {
    let wrapper: ShallowWrapper;
    let baseProps: Props;

    beforeEach(() => {
      baseProps = {
        answers: [create<Answer>('answer'), create<Answer>('answer')],
        totalCount: 2,
        questionState: QuestionState.Finished,
      };
    });

    describe('when question is not finished', () => {
      it('should render ErrorMessage', () => {
        const props = { ...baseProps, questionState: QuestionState.Answered };
        wrapper = shallow(<TrainingResult {...props} />);
        expect(wrapper.find(ErrorMessage).length).toBe(1);
      });
    });

    describe('when question is finished', () => {
      it('should render TrainingResultView', () => {
        wrapper = shallow(<TrainingResult {...baseProps} />);
        expect(wrapper.find(TrainingResultView).length).toBe(1);
      });

      it('should navigate to result when onClickRestart fired', () => {
        wrapper = shallow(<TrainingResult {...baseProps} />);
        wrapper
          .find(TrainingResultView)
          .props()
          .onClickRestart();
        expect(navigate).toHaveBeenCalled();
      });
    });
  });

  describe('mapStateToProps', () => {
    it('should convert state to props', () => {
      const state: GlobalState = {
        questions: {
          ...questionsState,
          answers: [create<Answer>('answer')],
          questionState: QuestionState.Finished,
        },
        ui: uiState,
      };

      expect(mapStateToProps(state)).toMatchSnapshot();
    });
  });
});
