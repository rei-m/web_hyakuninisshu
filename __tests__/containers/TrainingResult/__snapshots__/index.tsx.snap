// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrainingResult /> mapStateToProps should convert state to props 1`] = `
Object {
  "answers": Array [
    Object {
      "correct": true,
      "karutaNo": 1,
      "questionId": 1,
      "time": 1000000,
    },
  ],
  "questionState": 3,
  "totalCount": 0,
}
`;
